app/src/org/commcare/xml/AndroidCaseXmlParser.java

```diff
@@ -107,6 +109,46 @@ public void commit(Case parsed) throws IOException {
         }
     }
 
+    private void clearEntityCache(Case aCase) {
+        if (aCase == null || mEntityCache == null) {
+            return;
+        }
+        Set<Integer> recordsToWipe = new HashSet<>();
+        clearEntityCacheHelper(aCase, recordsToWipe, mCaseIndexTable, storage());
+        // Invalidate the current case in cache
+        mEntityCache.invalidateCaches(recordsToWipe);
+    }
+
+    public static void clearEntityCacheHelper(Case aCase, Set<Integer> visited, AndroidCaseIndexTable caseIndexTable,

```

Comment:
Instead of being recursive, I would update this methodology to use the [existing case index map](https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/models/database/user/models/AndroidCaseIndexTable.java#L95) functionality, similar to how we generate the DAG for [the case purge algorithm](https://github.com/dimagi/commcare-android/blob/2a031f9e206a92ee013b58a1b0880ebca88202cb/app/src/org/commcare/engine/cases/CaseUtils.java#L165), and then just run the purge in a loop with the list of cases. 

That would let us keep logic (and especially optimizations) for doing these kinds of case tree walks in one place, and keep the same guarantees about the shapes of the graphs.